src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -560,6 +563,150 @@ private async ValueTask SendRstStreamAsync(int streamId, Http2ProtocolErrorCode
+            {
+                authority = request.Headers.Host;
+            }
+            else
+            {
+                authority = request.RequestUri.IdnHost;
+                if (!request.RequestUri.IsDefaultPort)
+                {
+                    // TODO: Avoid allocation here by caching this on the connection or pool
+                    authority += ":" + request.RequestUri.Port;
+                }
+            }
+
+            WriteHeader(":authority", authority);
+            WriteHeader(":path", request.RequestUri.GetComponents(UriComponents.PathAndQuery | UriComponents.Fragment, UriFormat.UriEscaped));
+
+            if (request.HasHeaders)
+            {
+                WriteHeaders(request.Headers);
+            }
+
+            // Determine cookies to send.
+            if (_pool.Settings._useCookies)
+            {
+                string cookiesFromContainer = _pool.Settings._cookieContainer.GetCookieHeader(request.RequestUri);
+                if (cookiesFromContainer != string.Empty)
+                {
+                    WriteHeader(HttpKnownHeaderNames.Cookie, cookiesFromContainer);
+                }
+            }
+
+            if (request.Content == null)
+            {
+                // Write out Content-Length: 0 header to indicate no body,
+                // unless this is a method that never has a body.
+                if (normalizedMethod.MustHaveRequestBody)
+                {
+                    // TODO: ISSUE 31307: Use static table for Content-Length
+                    WriteHeader("Content-Length", "0");
+                }
+            }
+            else
+            {
+                WriteHeaders(request.Content.Headers);
+            }
+        }
+
+        private async ValueTask SendHeadersAsync(int streamId, HttpRequestMessage request)
+        {
+            // Note, HEADERS and CONTINUATION frames must be together, so hold the writer lock across sending all of them.
+            // We also serialize usage of the header encoder and the header buffer this way.
+            await _writerLock.WaitAsync().ConfigureAwait(false);
+
+            // Generate the entire header block, without framing, into the connection header buffer.
+            WriteHeaders(request);
+
+            ReadOnlyMemory<byte> remaining = _headerBuffer.ActiveMemory;
+            Debug.Assert(remaining.Length > 0);
+
+            // Split into frames and send.
+            try
+            {
+                ReadOnlyMemory<byte> current;
+                (current, remaining) = SplitBufferForFraming(remaining);
+
+                FrameFlags flags =
+                    (remaining.Length == 0 ? FrameFlags.EndHeaders : FrameFlags.None) |
+                    (request.Content == null ? FrameFlags.EndStream : FrameFlags.None);
+
+                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size + current.Length);
+                WriteFrameHeader(new FrameHeader(current.Length, FrameType.Headers, flags, streamId));
```

Comment:
As @caesar1995 mentioned in your other PR, there is some duplication of work going on here. `WriteFrameHeader` already calls `EnsureAvailableSpace(FrameHeader.Size)`. Since most frames have additional data and have to call `EnsureAvailableSpace` either way, I think we should probably remove that the call `WriteFrameHeader`. This isn't a huge amount of duplicate work though, so I'm ok with keeping it if there is a compelling reason.

Reply:
I agree, this should probably be removed from WriteFrameHeader. I'll make this change.